### Importing neccessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import string

In [2]:
import time

In [3]:
import random

Importing the data into the notebook using pandas

In [4]:
data = pd.read_csv('PasswordData.csv',low_memory=False)

#### Overview of data

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 669879 entries, 0 to 669878
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   password    669878 non-null  object 
 1   strength    669879 non-null  object 
 2   Unnamed: 2  238 non-null     float64
 3   Unnamed: 3  1 non-null       object 
 4   Unnamed: 4  1 non-null       float64
dtypes: float64(2), object(3)
memory usage: 25.6+ MB


In [6]:
data.head()

password strength  Unnamed: 2 Unnamed: 3  Unnamed: 4
0     kzde5577        1         NaN        NaN         NaN
1     kino3434        1         NaN        NaN         NaN
2    visi7k1yr        1         NaN        NaN         NaN
3     megzy123        1         NaN        NaN         NaN
4  lamborghin1        1         NaN        NaN         NaN

creating a copy of original data

In [7]:
datatable = data.copy()

Checking for the null values in data

In [8]:
datatable.notnull().sum()

password      669878
strength      669879
Unnamed: 2       238
Unnamed: 3         1
Unnamed: 4         1
dtype: int64

#### Data cleaning

- We see that their are some Unnamed columns which are nothing but impurities in data(culumns created by extra values).
- Point to observe is that these are not missing or null values that can be filled.
- They are also affecting the strength column as their data type is object instead of int of float
- Thus they need to be removed.
- let's cleane the data remove the columns and the corresponding affected rows

In [9]:
datatable.drop(['Unnamed: 3','Unnamed: 4'],axis=1,inplace=True)

In [10]:
datatable.head()

password strength  Unnamed: 2
0     kzde5577        1         NaN
1     kino3434        1         NaN
2    visi7k1yr        1         NaN
3     megzy123        1         NaN
4  lamborghin1        1         NaN

In [11]:
datatable.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 669879 entries, 0 to 669878
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   password    669878 non-null  object 
 1   strength    669879 non-null  object 
 2   Unnamed: 2  238 non-null     float64
dtypes: float64(1), object(2)
memory usage: 15.3+ MB


let's delete the Unnamed 2 column and corresponding corrupted 238
rows in the original data

In [12]:
datatable.dtypes

password       object
strength       object
Unnamed: 2    float64
dtype: object

Selecting only that part of data which contains 'NaN' in 'Unnamed: 2'

In [13]:
passdata = datatable[datatable['Unnamed: 2'].isnull() == True]

In [14]:
passdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669641 entries, 0 to 669878
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   password    669640 non-null  object 
 1   strength    669641 non-null  object 
 2   Unnamed: 2  0 non-null       float64
dtypes: float64(1), object(2)
memory usage: 20.4+ MB


In [15]:
passdata.drop(['Unnamed: 2'],axis = 1, inplace = True)

C:\Users\Aman\AppData\Local\Temp\ipykernel_15984\2967245833.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  passdata.drop(['Unnamed: 2'],axis = 1, inplace = True)


In [16]:
passdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669641 entries, 0 to 669878
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   password  669640 non-null  object
 1   strength  669641 non-null  object
dtypes: object(2)
memory usage: 15.3+ MB


There are still some row(s) that contains the object datatype in strength column
and which is hindering conversion of strength datatype to 'int64'.
    We need to remove it.

In [17]:
passdata = passdata[~(passdata['strength'].str.contains('.com'))]

Removed all the rows that contains .com type data in strength column

In [18]:
passdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669640 entries, 0 to 669878
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   password  669639 non-null  object
 1   strength  669640 non-null  object
dtypes: object(2)
memory usage: 15.3+ MB


Converting strength into int type

In [19]:
passdata['strength'] = passdata['strength'].astype('int64')

In [20]:
passdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669640 entries, 0 to 669878
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   password  669639 non-null  object
 1   strength  669640 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 15.3+ MB


In [21]:
passdata = passdata[passdata['password'].isnull() == False]

In [22]:
passdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669639 entries, 0 to 669878
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   password  669639 non-null  object
 1   strength  669639 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 15.3+ MB


In [23]:
orgdata = passdata.copy()

In [24]:
orgdata.head()

password  strength
0     kzde5577         1
1     kino3434         1
2    visi7k1yr         1
3     megzy123         1
4  lamborghin1         1

In [25]:
orgdata['strength'] = orgdata['strength'].astype('category')

In [26]:
orgdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669639 entries, 0 to 669878
Data columns (total 2 columns):
 #   Column    Non-Null Count   Dtype   
---  ------    --------------   -----   
 0   password  669639 non-null  object  
 1   strength  669639 non-null  category
dtypes: category(1), object(1)
memory usage: 10.9+ MB


In [27]:
orgdata.head()

password strength
0     kzde5577        1
1     kino3434        1
2    visi7k1yr        1
3     megzy123        1
4  lamborghin1        1

In [28]:
moderate, weak, strong =orgdata['strength'].value_counts()

In [29]:
print("Weak Passwords: ",weak)
print("Moderate Passwords: ",moderate)
print("Strong Passwords: ",strong)

Weak Passwords:  89701
Moderate Passwords:  496801
Strong Passwords:  83137


In [30]:
allpass = {"Moderate":moderate, "Weak":weak, "Strong":strong}

In [31]:
#total number of rows in the dataset
rows = len(orgdata)

In [32]:
#percentage of each type of class in data
for size in allpass.items():
    print("{}".format(size[0]),(size[1]/rows)*100)

Moderate 74.18937666414291
Weak 13.39542649098992
Strong 12.415196844867161


In [33]:
major_class = orgdata[orgdata["strength"] == 1]
minor_class0 = orgdata[orgdata["strength"] == 0]
minor_class2 = orgdata[orgdata["strength"] == 2]

In [34]:
major_class

password strength
0           kzde5577        1
1           kino3434        1
2          visi7k1yr        1
3           megzy123        1
4        lamborghin1        1
...              ...      ...
669874    10redtux10        1
669875     infrared1        1
669876  184520socram        1
669877     marken22a        1
669878      fxx4pw4g        1

[496801 rows x 2 columns]

In [35]:
# random sample size for major class

sample_size = random.randint(len(minor_class2),len(minor_class0))
sample_size

88641

In [36]:
# Assuming 'sample_size' is the desired size of the majority class
major_class_down = major_class.sample(n=sample_size, random_state=42)

In [37]:
major_class_down

password strength
459487    martha1968        1
195076  pablo%varela        1
514888     Y79917042        1
421758  yamahaerg121        1
111088   loveless025        1
...              ...      ...
619956    gorkemm123        1
175468    seetos4439        1
300297  8240kushwaha        1
373253    gmb7014712        1
90755      xsl745836        1

[88641 rows x 2 columns]

In [38]:
# concatenating the all three classes togather
orgdata = pd.concat([minor_class0,minor_class2,major_class_down], ignore_index = True)

In [39]:
orgdata

password strength
0             intel1        0
1            lsdlsd1        0
2             olmaz.        0
3             ok>bdk        0
4             d04m11        0
...              ...      ...
261474    gorkemm123        1
261475    seetos4439        1
261476  8240kushwaha        1
261477    gmb7014712        1
261478     xsl745836        1

[261479 rows x 2 columns]

In [40]:
orgdata = orgdata.sample(frac=1, random_state=42).reset_index(drop=True)
orgdata

password strength
0                leonre4        0
1           paulonunes18        1
2                zillya1        0
3                gsahen1        0
4       bG3manTU0NwKj7d0        2
...                  ...      ...
261474       iswlztnjp76        1
261475  9JYleIDg4MQq91FK        2
261476  17611994MaTT8904        2
261477  V4KJRHzA4MgA86GK        2
261478  OBVhGnTc1MAWnis1        2

[261479 rows x 2 columns]

In [41]:
orgdata['strength'] = orgdata['strength'].map({0:'Weak',
                                               1:'Moderate',
                                               2:'Strong'})

In [42]:
orgdata

password  strength
0                leonre4      Weak
1           paulonunes18  Moderate
2                zillya1      Weak
3                gsahen1      Weak
4       bG3manTU0NwKj7d0    Strong
...                  ...       ...
261474       iswlztnjp76  Moderate
261475  9JYleIDg4MQq91FK    Strong
261476  17611994MaTT8904    Strong
261477  V4KJRHzA4MgA86GK    Strong
261478  OBVhGnTc1MAWnis1    Strong

[261479 rows x 2 columns]

### Tokenizing and Trainning a Machine learning model

In [43]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

In [44]:
def word(password):
    character=[]
    for i in password:
        character.append(i)
    return character

In [45]:
x = np.array(orgdata["password"])
y = np.array(orgdata["strength"])

In [46]:
tdif = TfidfVectorizer(tokenizer=word)
x = tdif.fit_transform(x)
xtrain, xtest, ytrain, ytest = train_test_split(x, y, 
                                                test_size=0.05, 
                                                random_state=42)

C:\Anaconda_Install\lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [47]:
start_time = time.time()

model = RandomForestClassifier()
model.fit(xtrain, ytrain)
print(model.score(xtest, ytest))

end_time = time.time()
elapsed_time = end_time - start_time

# Print the elapsed time
print("Elapsed time:", elapsed_time, "seconds")

0.9490592014685636
Elapsed time: 1396.8695740699768 seconds


In [51]:
userpass = input("Enter Password: ")
data = tdif.transform([userpass]).toarray()
output = model.predict(data)
print(output)


Enter Password: fhvhjg7iui98
['Strong']
